cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/RemoteTaskExecutor.java

```diff
@@ -80,35 +90,87 @@ public RemoteTaskExecutor(CConfiguration cConf, RemoteClientFactory remoteClient
-          .withBody(requestBody.duplicate());
-        if (compression) {
-          requestBuilder.addHeader(HttpHeaders.CONTENT_ENCODING, "gzip");
-          requestBuilder.addHeader(HttpHeaders.ACCEPT_ENCODING, "gzip, deflate");
-        }
-        HttpRequest httpRequest = requestBuilder.build();
+    try {
+      RetryableTaskResult retryableTaskResult = Retries.callWithRetries((attempt) -> {
+        try {
+          HttpRequest.Builder requestBuilder = remoteClient
+            .requestBuilder(HttpMethod.POST, TASK_WORKER_URL)
+            .withBody(requestBody.duplicate());
+          if (compression) {
+            requestBuilder.addHeader(HttpHeaders.CONTENT_ENCODING, "gzip");
+            requestBuilder.addHeader(HttpHeaders.ACCEPT_ENCODING, "gzip, deflate");
+          }
+          HttpRequest httpRequest = requestBuilder.build();
 
-        HttpResponse httpResponse = remoteClient.execute(httpRequest);
-        if (httpResponse.getResponseCode() == HttpResponseStatus.TOO_MANY_REQUESTS.code()) {
+          HttpResponse httpResponse = remoteClient.execute(httpRequest);
+          if (httpResponse.getResponseCode() == HttpResponseStatus.TOO_MANY_REQUESTS.code()) {
+            throw new RetryableException(
+              String.format("Received response code %s for %s", httpResponse.getResponseCode(),
+                            runnableTaskRequest.getClassName()));
+          }
+          if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
+            BasicThrowable basicThrowable = GSON
+              .fromJson(new String(getResponseBody(httpResponse)), BasicThrowable.class);
+            throw RemoteExecutionException.fromBasicThrowable(basicThrowable);
+          }
+          return new RetryableTaskResult(getResponseBody(httpResponse), attempt);
+        } catch (NoRouteToHostException e) {
           throw new RetryableException(
-            String.format("Received response code %s for %s", httpResponse.getResponseCode(),
-                          runnableTaskRequest.getClassName()));
-        }
-        if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
-          BasicThrowable basicThrowable = GSON
-            .fromJson(new String(getResponseBody(httpResponse)), BasicThrowable.class);
-          throw RemoteExecutionException.fromBasicThrowable(basicThrowable);
+            String.format("Received exception %s for %s", e.getMessage(), runnableTaskRequest.getClassName()));
         }
-        return getResponseBody(httpResponse);
-      } catch (NoRouteToHostException e) {
-        throw new RetryableException(
-          String.format("Received exception %s for %s", e.getMessage(), runnableTaskRequest.getClassName()));
+      }, retryStrategy, Retries.DEFAULT_PREDICATE);
+      //emit metrics with successful result
+      emitMetrics(startTime, true, runnableTaskRequest, retryableTaskResult.getAttempts());
```

Comment:
This actually can be done inside the lambda so that you don't need the new class RetryableTaskResult. You can do it right before the `return` line (line 120 in the current code).

Reply:
Done. Removed RetryableTaskResult class.